File: backend/app/schemas/subscription.py
import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.models.payment import PaymentPurpose, PaymentStatus
from app.models.subscription import BillingInterval, SubscriptionStatus


class SubscriptionPlanRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    code: str
    name: str
    billing_interval: BillingInterval
    price: Decimal
    trial_days: int


class SubscriptionCreate(BaseModel):
    plan_code: str


class SubscriptionRead(BaseModel):
    id: uuid.UUID
    plan: SubscriptionPlanRead
    status: SubscriptionStatus
    current_period_start: datetime
    current_period_end: datetime
    cancel_at_period_end: bool
    cancelled_at: datetime | None


class CheckoutResponse(BaseModel):
    payment_id: uuid.UUID
    checkout_url: str | None
    status: str


class PaymentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    purpose: PaymentPurpose
    amount: Decimal
    status: PaymentStatus
    gateway: str | None
    paid_at: datetime | None
    created_at: datetime

File: frontend/src/types/subscription.ts
export type BillingInterval = "MONTHLY" | "ANNUAL";

export type SubscriptionStatus = "TRIAL" | "ACTIVE" | "PAST_DUE" | "CANCELLED" | "EXPIRED";

export type PaymentStatus = "PENDING" | "PAID" | "FAILED" | "REFUNDED";

export type PaymentPurpose = "SUBSCRIPTION" | "APPOINTMENT";

export interface SubscriptionPlan {
  id: string;
  code: string;
  name: string;
  billing_interval: BillingInterval;
  price: string;
  trial_days: number;
}

export interface Subscription {
  id: string;
  plan: SubscriptionPlan;
  status: SubscriptionStatus;
  current_period_start: string;
  current_period_end: string;
  cancel_at_period_end: boolean;
  cancelled_at: string | null;
}

export interface CheckoutResponse {
  payment_id: string;
  checkout_url: string | null;
  status: string;
}

export interface SubscriptionPayment {
  id: string;
  purpose: PaymentPurpose;
  amount: string;
  status: PaymentStatus;
  gateway: string | null;
  paid_at: string | null;
  created_at: string;
}
